src/swap/saga.ts

```diff
@@ -0,0 +1,116 @@
+import { CeloTx } from '@celo/connect'
+import { TxParamsNormalizer } from '@celo/connect/lib/utils/tx-params-normalizer'
+import { ContractKit } from '@celo/contractkit'
+import { PayloadAction } from '@reduxjs/toolkit'
+import { call, put, select, takeLatest } from 'redux-saga/effects'
+import { maxSwapSlippagePercentageSelector } from 'src/app/selectors'
+import { navigate } from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import {
+  swapApprove,
+  swapError,
+  swapExecute,
+  swapPriceChange,
+  swapStart,
+  swapSuccess,
+} from 'src/swap/slice'
+import { Field, SwapInfo, SwapTransaction } from 'src/swap/types'
+import { sendTransaction } from 'src/transactions/send'
+import { newTransactionContext } from 'src/transactions/types'
+import { fetchWithTimeout } from 'src/utils/fetchWithTimeout'
+import Logger from 'src/utils/Logger'
+import { getContractKit } from 'src/web3/contracts'
+import networkConfig from 'src/web3/networkConfig'
+import { getConnectedUnlockedAccount } from 'src/web3/saga'
+import { applyChainIdWorkaround, buildTxo } from 'src/web3/utils'
+
+const TAG = 'swap/saga'
+
+function getPriceDiff(price1: number, price2: number) {
+  return (Math.abs(price1 - price2) / ((price1 + price2) / 2)) * 100
+}
+
+export function* swapSubmitSaga(action: PayloadAction<SwapInfo>) {
```

Comment:
i am wondering if there is value in this `SwapInfo` object, since it doesn't occur naturally and no component / function seems to require all the properties at once? for example, the payload for this action could be something like `{approveTransaction, unvalidatedSwapTransaction, updatedFieldType}`. the benefit would be that you can make every one of these properties  required. currently in `SwapInfo`, the approveTransaction property is optional which is actually not expected for this saga or else line 58 will not have any tx object to approve - typescript didn't help to catch this in this case though...

Reply:
Does it makes sense to change the valora-rest-api to return a swapTransaction for both the approve and execute endpoints instead of unvalidatedSwapTransaction  or validatedSwapTransaction these appear to have all the same fields.

Reply:
I've opened https://github.com/valora-inc/valora-rest-api/pull/162 to change the api behavior to above.